[WPE][GTK] Fix leaks when using g_variant_iter_loop()

mcatanzaro opened the pull request:
#### 449035bdfe62c6434ed30eb8df64ecd92f1b76f0
<pre>
[WPE][GTK] Fix leaks when using g_variant_iter_loop()
<a href="https://bugs.webkit.org/show_bug.cgi?id=298066">https://bugs.webkit.org/show_bug.cgi?id=298066</a>

Reviewed by Patrick Griffis.

Milan pointed out that if we break out of a g_variant_iter_loop() loop,
then it cannot free the key or value for us. In these cases, we have to
free it manually.

The easiest solution is to just not use g_variant_iter_loop() in these
cases, to ensure we cannot mess up. g_variant_iter_next() is basically
the same thing, except it requires that we handle the ownership.

* Source/WebKit/Shared/glib/JavaScriptEvaluationResultGLib.cpp:
(WebKit::JavaScriptEvaluationResult::GLibExtractor::toValue):
(WebKit::isSerializable):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(parseAsyncFunctionArguments):

Canonical link: <a href="https://commits.webkit.org/315630@main">https://commits.webkit.org/315630@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/20d647acd77b04f817823fa6699acba83db4528f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/169602 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/43524 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/36868 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/178961 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/127314 "Built successfully") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/44532 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/43153 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/132151 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/93079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/172561 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/35029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/152511 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/112654 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/155/builds/44532 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/32288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/27658 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/144001 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/31910 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/181560 "Built successfully") | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/34268 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/36454 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/140600 "Passed tests") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/43180 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/140436 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/37790 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/43869 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/170/builds/28890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/108092 "Built successfully") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/35701 "Passed tests") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/115634 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/42379 "Built successfully") | [✅ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/185/builds/6975 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/41772 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/42027 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/41915 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro committed:
[WPE][GTK] Fix leaks when using g_variant_iter_loop()
https://bugs.webkit.org/show_bug.cgi?id=298066

Reviewed by NOBODY (OOPS!).

Milan pointed out that if we break out of a g_variant_iter_loop() loop,
then it cannot free the key or value for us. In these cases, we have to
free it manually.

carlosgcampos commented on Source/WebKit/Shared/glib/JavaScriptEvaluationResultGLib.cpp:
```diff
@@ -121,8 +121,12 @@ static bool isSerializable(GVariant* variant)
         GVariant* value;
         while (g_variant_iter_loop(&iter, "{&sv}", &key, &value)) {
-            if (!key || !isSerializable(value))
+            if (!key || !isSerializable(value)) {
+                // Normally g_variant_iter_loop() frees this for us, but not if
+                // we break out of the loop and never call it again.
+                g_variant_unref(value);
```
I think it would be clearer to use GRefPtr for the GVariant and g_variant_iter_next in this case. Note there's another case in this file. Since we have smart pointers for GVariant maybe it's less risky and confusing to always use g_variant_iter_next in webkit, but I'm ok if we just change the cases where we are leaking.

TingPing reviewed (approved):
I find relying on `outPtr` in the loop to be a bit non-obvious, but looks correct.

mcatanzaro committed:
[WPE][GTK] Fix leaks when using g_variant_iter_loop()
https://bugs.webkit.org/show_bug.cgi?id=298066

Reviewed by Patrick Griffis.

Milan pointed out that if we break out of a g_variant_iter_loop() loop,
then it cannot free the key or value for us. In these cases, we have to
free it manually.

The easiest solution is to just not use g_variant_iter_loop() in these
cases, to ensure we cannot mess up. g_variant_iter_next() is basically
the same thing, except it requires that we handle the ownership.

* Source/WebKit/Shared/glib/JavaScriptEvaluationResultGLib.cpp:
(WebKit::JavaScriptEvaluationResult::GLibExtractor::toValue):
(WebKit::isSerializable):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(parseAsyncFunctionArguments):

Canonical link: https://commits.webkit.org/315630@main
